### PR TITLE
Renamings

### PR DESCRIPTION
**Pull request by jtakalai**
Fixed Delegation object (and associated handlers) in subgraph

Lots of renamings, among others:
   * PoolValueUpdate => OperatorValueUpdate
   * PoolYieldPolicy => ExchangeRatePolicy
   * pooltokenToData => undelegationRate
   * dataToPooltoken => delegationRate
   * event Profit(uint valueIncreaseWei, uint operatorsCutDataWei, uint protocolFeeDataWei);
   * event Loss(uint valueDecreaseWei);
   * amountPoolTokenWei => amountOperatorTokenWei
   * poolTokenName => operatorTokenName
   * _mintPoolTokensFor => _delegate
   * rewardThreshold => maxAllowedEarnings
   * poolValueDriftLimitFraction => maxAllowedEarningsFraction
   * poolValueDriftPenaltyFraction => fishermanRewardFraction
   * approximatePoolValue => operatorValueWithoutEarnings
   * poolValue => valueWithoutEarnings
   * poolValueBlockNumber, poolValueTimestamp => valueUpdateTimestamp/BlockNumber
   * sponsorship.unallocatedWei => sponsorship.remainingWei
   * sponsorship.allocatedWei => sponsorship.earningsWei
   * committed stake => locked stake

### CODE REVIEW

File: packages/network-contracts/contracts/OperatorTokenomics/Sponsorship.sol [hunk 1]
```diff
@@ -75,8 +72,8 @@ contract Sponsorship is Initializable, ERC2771ContextUpgradeable, IERC677Receive
 
     mapping(address => uint) public stakedWei; // how much each operator has staked, if 0 operator is considered not part of sponsorship
     mapping(address => uint) public joinTimeOfOperator;
-    mapping(address => uint) public committedStakeWei; // how much can not be unstaked (during e.g. flagging)
-    uint public committedForfeitedStakeWei; // committedStakeWei that has been forfeited but still needs to be tracked to e.g. pay the flag reviewers
+    mapping(address => uint) public lockedStakeWei; // how much can not be unstaked (during e.g. flagging)
+    uint public lockedForfeitedStakeWei; // lockedStakeWei that has been forfeited but still needs to be tracked to e.g. pay the flag reviewers
```
Comment (teogeb) [inline packages/network-contracts/contracts/OperatorTokenomics/Sponsorship.sol:76]:
```suggestion
    uint public forfeitedStakeWei; // lockedStakeWei that has been forfeited but still needs to be tracked to e.g. pay the flag reviewers
```

no need to prefix as there is only one type of forfeited stake (if we "unlock" some of the forfeited stake then it is no longer forfeited)

File: packages/network-contracts/contracts/OperatorTokenomics/StreamrConfig.sol [hunk 2]
```diff
@@ -48,29 +48,27 @@ contract StreamrConfig is Initializable, UUPSUpgradeable, AccessControlUpgradeab
...
      */
     uint public maxPenaltyPeriodSeconds;
 
     /**
-     * The real-time precise pool value can not be kept track of, since it would mean looping through all sponsorships in each transaction.
-     * Everyone can update the "pool-value" of a list of Sponsorships.
-     * If the the withdrawn earnings are more than poolValueDriftLimitFraction * (stake + free funds before withdraw),
-     *   part of the operator's cut is sent to the withdrawEarningsFromSponsorships caller, which can be anyone.
-     * This means operator should call withdrawEarningsFromSponsorships often enough to not accumulate too much unwithdrawn earnings, so they can keep all of the cut.
+     * The real-time precise operator value can not be kept track of, since it would mean looping through all Sponsorships in each transaction.
```
Comment (teogeb) [inline packages/network-contracts/contracts/OperatorTokenomics/StreamrConfig.sol:55]: Was there a exact definition of the real-time operator value somewhere in the comments? Was the big picture like this:
- real operator value is stakes + profits + earnings (+ something?)
- when earnings are withdrawn, earnings are converted to profits -> there are less earnings -> `valueWithoutEarnings` is then of course closer to "valueWithEarnings" (i.e. the real value) 
- if we'd earnings the real value and `valueWithoutEarnings` would be exactly same? (until more earnings are generated in sponsorships)

File: packages/network-contracts/contracts/OperatorTokenomics/Operator.sol [hunk 3]
```diff
@@ -30,28 +30,28 @@ import "../StreamRegistry/IStreamRegistryV4.sol";
 
 /**
  * Operator contract receives and holds the delegators' tokens.
- * The operator (owner()) stakes them to Sponsorships of the streams that the operator's nodes relay.
+ * The operator (`owner()`) stakes them to Sponsorships of the streams that the operator's nodes relay.
  * Operator contract also is an ERC20 token that each delegator receives, and can swap back to DATA when they undelegate.
  *
- * @dev DATA token balance of the pool === the "free funds" available for staking,
+ * @dev DATA token balance of the contract === the "free funds" available for staking,
```
Comment (teogeb) [inline packages/network-contracts/contracts/OperatorTokenomics/Operator.sol:36]: Would it make sense to rename free funds to `nonstakedAmount`? At least for me the definition was not very clear until I read this comment.

File: packages/network-contracts/scripts/tatum/StreamrEnvDeployerHardhat.ts [hunk 4]
```diff
@@ -28,7 +28,7 @@ export type EnvContracAddresses = {
     "StreamrConfig": string,
     "SponsorshipFactory": string,
     "SponsorshipDefaultDelegationPolicy": string,
-    // "SponsorshipDefaultPoolYieldPolicy": string,
+    // "SponsorshipDefaultExchangeRatePolicy": string,
```
Comment (teogeb) [inline packages/network-contracts/scripts/tatum/StreamrEnvDeployerHardhat.ts:31]:
```suggestion
```

File: packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts [hunk 5]
```diff
@@ -294,26 +294,26 @@ describe("Operator contract", (): void => {
             await advanceToTimestamp(timeAtStart, "Stake to sponsorship")
             await (await operator.stake(sponsorship.address, parseEther("1000"))).wait()
 
-            // some time passes => approx poolvalue differs from real poolvalue
+            // some time passes => approx valueWithoutEarnings differs from real valueWithoutEarnings
```
Comment (teogeb) [inline packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts:297]:
```suggestion
            // some time passes => valueWithoutEarnings differs from real operator value
```

File: packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts [hunk 6]
```diff
@@ -872,8 +872,8 @@ describe("Operator contract", (): void => {
         expect(await operator.balanceOf(delegator.address)).to.equal(parseEther("50"))
 
         // now anyone can trigger the unstake and payout of the queue
-        // await (await operator.updateApproximatePoolvalueOfSponsorship(sponsorship2.address)).wait()
-        // await (await operator.updateApproximatePoolvalueOfSponsorship(sponsorship1.address)).wait()
+        // await (await operator.updateApproximatevalueOfSponsorship(sponsorship2.address)).wait()
+        // await (await operator.updateApproximatevalueOfSponsorship(sponsorship1.address)).wait()
```
Comment (teogeb) [inline packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts:876]: What is the up-to-date name of this function?

**Review by teogeb — COMMENTED** [review]
Great improvements! 👍 👍

File: packages/network-contracts/contracts/OperatorTokenomics/Operator.sol [hunk 7]
```diff
@@ -30,28 +30,28 @@ import "../StreamRegistry/IStreamRegistryV4.sol";
...
+    event QueueUpdated(address delegator, uint amountWei, uint queueIndex);
 
     // sponsorship events (initiated by CONTROLLER_ROLE)
     event Staked(Sponsorship indexed sponsorship);
     event Unstaked(Sponsorship indexed sponsorship);
     event StakeUpdate(Sponsorship indexed sponsorship, uint stakedWei);
-    event PoolValueUpdate(uint totalStakeInSponsorshipsWei, uint freeFundsWei); // DATA token tracking event (staked - slashed)
-    event Profit(uint poolIncreaseWei, uint operatorsCutDataWei, uint protocolFeeDataWei);
-    event Loss(uint poolDecreaseWei);
+    event OperatorValueUpdate(uint totalStakeInSponsorshipsWei, uint freeFundsWei); // DATA token tracking event (staked - slashed)
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply packages/network-contracts/contracts/OperatorTokenomics/Operator.sol:52]: note that stake = stakeD - slashed. So it's probably correct.

File: packages/network-contracts/contracts/OperatorTokenomics/Operator.sol [hunk 8]
```diff
@@ -315,18 +315,23 @@ contract Operator is Initializable, ERC2771ContextUpgradeable, IERC677Receiver,
...
-     * @param amountPoolTokenWei amount of operator tokens to convert back to DATA. Can be more than the balance; then all operator tokens are undelegated.
+     * Add the request to undelegate into the undelegation queue. When new earnings arrive, they will be used to pay out the queue in order.
+     * It's all call `undelegate` with any `amountWei`, the actual amount is decided when it's your turn, and will be capped to the actual balance at the time.
+     * NOTE: "Undelegate all" request can be made by calling this function e.g. like so: `operator.undelegate(maxUint256)`,
+     *       where `maxUint256 = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF` (or `2**256 - 1`).
+     * @param amountWei of operator tokens to convert back to DATA. Can be more than the balance; then all operator tokens are undelegated.
      **/
-    function undelegate(uint amountPoolTokenWei) public {
-        moduleCall(address(queueModule), abi.encodeWithSelector(queueModule._undelegate.selector, amountPoolTokenWei));
+    function undelegate(uint amountWei) public {
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply packages/network-contracts/contracts/OperatorTokenomics/Operator.sol:328]: you could say :)

File: packages/network-contracts/contracts/OperatorTokenomics/OperatorPolicies/QueueModule.sol [hunk 9]
```diff
@@ -49,26 +49,25 @@ contract QueueModule is IQueueModule, Operator {
...
             delete undelegationQueue[queueCurrentIndex];
             emit QueueUpdated(delegator, 0, queueCurrentIndex);
             queueCurrentIndex++;
             return 0;
         }
 
         // convert to DATA and see if we have enough free funds to pay out the queue item in full
-        uint amountDataWei = moduleCall(address(yieldPolicy), abi.encodeWithSelector(yieldPolicy.pooltokenToData.selector,
-            amountPoolTokens, 0));
+        uint amountDataWei = moduleCall(address(exchangeRatePolicy), abi.encodeWithSelector(exchangeRatePolicy.undelegationRate.selector, amountOperatorTokens));
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply packages/network-contracts/contracts/OperatorTokenomics/OperatorPolicies/QueueModule.sol:70]: not needed. It was always set to zero. It's an adjustment that is needed when delegating.

File: packages/network-contracts/scripts/tatum/StreamrEnvDeployerHardhat.ts (hunk 4 above)
Comment (jtakalai) [reply packages/network-contracts/scripts/tatum/StreamrEnvDeployerHardhat.ts:31]: (other than the obvious renaming)

File: packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts (hunk 6 above)
Comment (jtakalai) [reply packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts:876]: `withdrawEarningsFromSponsorships` I think

File: packages/network-contracts/contracts/OperatorTokenomics/Operator.sol (hunk 3 above)
Comment (jtakalai) [reply packages/network-contracts/contracts/OperatorTokenomics/Operator.sol:36]: updated to `dataTokenBalance`

File: packages/network-contracts/contracts/OperatorTokenomics/OperatorPolicies/DefaultExchangeRatePolicy.sol [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import "../Operator.sol";
+
+contract DefaultExchangeRatePolicy is IExchangeRatePolicy, Operator {
+
+    function setParam(uint) external {
+
+    }
+
+    /**
+     * Exchange rate from Operator's internal token to DATA when undelegating
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply packages/network-contracts/contracts/OperatorTokenomics/OperatorPolicies/DefaultExchangeRatePolicy.sol]: reverted back to old names (~ish)

File: packages/network-contracts/contracts/OperatorTokenomics/OperatorPolicies/QueueModule.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (teogeb) [reply packages/network-contracts/contracts/OperatorTokenomics/OperatorPolicies/QueueModule.sol]: Could remove it from line 80 then?

File: packages/network-subgraphs/src/operator.ts [hunk 11]
```diff
@@ -123,53 +123,35 @@ export function handlePoolValueUpdate(event: PoolValueUpdate): void {
...
-//     }
-//     stake.date = event.block.timestamp
-//     stake.amount = event.params.amountWei
-//     stake.save()
-// }
-
 export function handleQueuedDataPayout(event: QueuedDataPayout): void {
     let operatorContractAddress = event.address.toHexString()
-    let amountPT = event.params.amountPoolTokenWei
+    let amountPT = event.params.amountWei
```
Comment (samigabor) [inline packages/network-subgraphs/src/operator.ts:154]: I guess would be amountOP?

File: packages/network-subgraphs/src/operator.ts [hunk 12]
```diff
@@ -46,22 +46,34 @@ export function handleBalanceUpdate(event: BalanceUpdate): void {
...
+
+/** Track how much has been undelegated (in DATA) */
+export function handleUndelegated(event: Undelegated): void {
+    let operatorContractAddress = event.address.toHexString()
+    let amountUndelegatedWei = event.params.amountDataWei
+    log.info('handleUndelegated: operatorContractaddress={} blockNumber={}', [operatorContractAddress, event.block.number.toString()])
+    log.info('handleUndelegated: amountDataWei={}', [amountUndelegatedWei.toString()])
+
+    let delegation = loadOrCreateDelegation(operatorContractAddress, event.params.delegator.toHexString(), event.block.timestamp)
+    delegation.undelegatedDataWei = delegation.undelegatedDataWei.plus(amountUndelegatedWei)
```
Comment (samigabor) [inline packages/network-subgraphs/src/operator.ts:73]: :+1:

**Review by samigabor — APPROVED** [review]

**Pull request merged**